### PR TITLE
Microsoft defender for cloud collector

### PR DESCRIPTION
**Pull request by omerKarkKatz**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-3992)
fixes: [link to the issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-4054)

## Description
This is a new XSIAM collector for Microsoft defender for cloud

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector_description.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+Azure security center
+| API / Permissions name | Type | Description |
+| --- | --- | --- |
+| Azure scurity center | --- | --- |
+| user_impresonation | Delegated | Access Azure security center |
+| SecurityCenter | --- | --- |
+| user_impresonation | Delegated | Access SecurityCenter |
+
+After you finish configuring your application with the proper permission.
```
Comment (ShirleyDenkberg) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector_description.md:19]:
```suggestion
After you finish configuring your application with the proper permission,
```

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector_description.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
+Microsoft Defender for Cloud Event Collector, provides unified security management and advanced threat protection across hybrid cloud workloads.
+This collector retrieves a list of all the alerts that are associated with a subscription.
+
+#### Authentication with Microsoft Defender For Cloud (Self deployed Azure App)
+
+self-deployed configuration:
```
Comment (ShirleyDenkberg) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector_description.md:6]:
```suggestion
Self-deployed configuration:
```

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector_description.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+Azure security center
+| API / Permissions name | Type | Description |
+| --- | --- | --- |
+| Azure scurity center | --- | --- |
+| user_impresonation | Delegated | Access Azure security center |
+| SecurityCenter | --- | --- |
+| user_impresonation | Delegated | Access SecurityCenter |
+
+After you finish configuring your application with the proper permission.
+Adda “Security Reader” role to the application from the subscription.
```
Comment (ShirleyDenkberg) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector_description.md:20]:
```suggestion
add a “Security Reader” role to the application from the subscription.
```

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector_description.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+In order to add a role to a subscription refer to.
```
Comment (ShirleyDenkberg) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector_description.md:21]:
```suggestion
In order to add a role to a subscription refer to:
```

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector_description.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+* [Azure AD built-in roles](https://learn.microsoft.com/en-us/azure/active-directory/roles/permissions-reference)
+
+#### Additional information
+
+For additional information on the API refer to.
```
Comment (ShirleyDenkberg) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector_description.md:26]:
```suggestion
For additional information on the API refer to:
```

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
+XSIAM collector for Microsoft Defender for Cloud alerts.
+This integration was integrated and tested with version xx of Microsoft Defender for Cloud Event Collector
+
+## Configure Microsoft Defender for Cloud Event Collector on Cortex XSOAR
```
Comment (ShirleyDenkberg) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md:4]:
```suggestion
## Configure Microsoft Defender for Cloud Event Collector on Cortex XSIAM
```

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+1. Navigate to **Settings** > **Integrations** > **Servers & Services**.
```
Comment (ShirleyDenkberg) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md:6]:
```suggestion
1. Navigate to **Settings** > **Configurations** > **Data Collection** > **Automations & Feed Integrations**.
```

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+2. Search for Microsoft Defender for Cloud Event Collector.
+3. Click **Add instance** to create and configure a new integration instance.
+
+    | **Parameter** | **Description** | **Required** |
+    | --- | --- | --- |
+    | Microsoft Azure Management URL |  | False |
+    | ID |  | True |
+    | Token |  | True |
+    | Key |  | True |
+    | Certificate Thumbprint | Used for certificate authentication. As appears in the "Certificates &amp;amp; secrets" page of the app. | False |
```
Comment (ShirleyDenkberg) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md:16]: Need to fix &amp;amp;
What is this supposed to be?

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    | Private Key | Used for certificate authentication. The private key of the registered certificate. | False |
+    | Subscription ID to use |  | True |
+    | Trust any certificate (not secure) |  | False |
+    | Use system proxy settings |  | False |
+
+4. Click **Test** to validate the URLs, token, and connection.
+
+## Commands
+
+You can execute these commands from the Cortex XSOAR CLI, as part of an automation, or in a playbook.
```
Comment (ShirleyDenkberg) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md:26]: Should this be the Cortex XSIAM CLI?

**Comment by ShirleyDenkberg** [comment]
@gal-forer @DeanArbel @yucohen Doc review completed.

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md (hunk 9 above)
Comment (omerKarkKatz) [reply Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md]: good catch changed it thanks

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md (hunk 8 above)
Comment (omerKarkKatz) [reply Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md]: fixed Thanks.

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector.py [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+import urllib3
+from MicrosoftApiModule import *  # noqa: E402
+# Disable insecure warnings
+urllib3.disable_warnings()
+
+''' CONSTANTS '''
+
+VENDOR = 'microsoft'
+PRODUCT = 'defender_for_cloud'
+API_VERSION = '2022-01-01'
```
Comment (gal-forer) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector.py:12]: Will this never  change for any client ?

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md [hunk 11]
```diff
@@ -0,0 +1,50 @@
+XSIAM collector for Microsoft Defender for Cloud alerts.
+This integration was integrated and tested with version xx of Microsoft Defender for Cloud Event Collector
```
Comment (gal-forer) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md:2]: Replace the xx with a real version

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    | --- | --- | --- |
+    | Microsoft Azure Management URL |  | False |
+    | ID |  | True |
+    | Token |  | True |
+    | Key |  | True |
+    | Certificate Thumbprint | Used for certificate authentication. As appears in the "Certificates & secrets" page of the app. | False |
+    | Private Key | Used for certificate authentication. The private key of the registered certificate. | False |
+    | Subscription ID to use |  | True |
+    | Trust any certificate (not secure) |  | False |
+    | Use system proxy settings |  | False |
```
Comment (gal-forer) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md:20]: Add missing description

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/test_data/AlertsToSort.json [hunk 13]
```diff
@@ -0,0 +1,175 @@
+[
+  {
+    "id": "/subscriptions/0f907ea4-bc8b-4c11-9d7e-805c2fd144fb/resourceGroups/Sample-RG/providers/Microsoft.Security/locations/centralus/alerts/2517396773228188817_67a23793-595b-4efb-ae23-679568966575",
```
Comment (gal-forer) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/test_data/AlertsToSort.json:3]: Is it a real ID ? Wit hthe subscription ?

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/test_data/AlertsToSort.json [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+          "location": {
+            "countryCode": "US",
+            "countryName": "United States",
+            "state": "Virginia",
+            "city": "Washington"
+          },
+          "type": "ip"
+        }
+      ],
+      "alertUri": "https://portal.azure.com/#blade/Microsoft_Azure_Security_AzureDefenderForData/AlertBlade/alertId/2517396773228188817_67a23793-595b-4efb-ae23-679568966575/subscriptionId/0f907ea4-bc8b-4c11-9d7e-805c2fd144fb/resourceGroup/Sample-RG/referencedFrom/alertDeepLink/location/centralus"
```
Comment (gal-forer) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/test_data/AlertsToSort.json:72]: Is the subscriptionId a secret ?

**Review by gal-forer — CHANGES_REQUESTED** [review]

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector.py (hunk 10 above)
Comment (omerKarkKatz) [reply Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector.py:12]: What do you mean by this? This is the latest API-version that we want to use.

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector.py [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+
+def check_events_were_filtered_out(events: list, filtered_events: list) -> bool:
+    """
+    Args:
+        - events (list): The events list response from the API
+        - filtered_events (list): The filtered event list (no dups)
+    Returns:
+        (bool): Whether events were filtered out
+    """
+    return len(events) > len(filtered_events)
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector.py:97]: Did this as a one-liner for UT.

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md (hunk 11 above)
Comment (omerKarkKatz) [reply Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md:2]: deleted this line.

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/test_data/AlertsToSort.json (hunk 13 above)
Comment (omerKarkKatz) [reply Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/test_data/AlertsToSort.json:3]: No its a Microsoft fake generated data

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector.py [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+        events.extend(curr_filtered_events)
+
+        while nextLink := response.get('nextLink', None):
+            response = self.ms_client.http_request(method="GET", full_url=nextLink)
+            curr_events = response.get("value", [])
+            curr_filtered_events = filter_out_previosly_digested_events(curr_events, last_run)
+            if check_events_were_filtered_out(curr_events, curr_filtered_events):
+                events.extend(curr_filtered_events)
+                break
+            events.extend(curr_filtered_events)
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector.py:68]: nice suggestion done.

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md (hunk 12 above)
Comment (omerKarkKatz) [reply Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md:20]: This is auto-generated from the yml added description where it is not self-explanatory.

**Review by gal-forer — APPROVED** [review]

**Review by DeanArbel — APPROVED** [review]
Looks good. Good job, I know it's been a bumpy road.
@ShirleyDenkberg Please review this PR, take special attention to the Known Limitation section.

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector_description.md [hunk 17]
```diff
@@ -0,0 +1,27 @@
+Microsoft Defender for Cloud Event Collector, provides unified security management and advanced threat protection across hybrid cloud workloads.
```
Comment (ShirleyDenkberg) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector_description.md:1]:
```suggestion
Microsoft Defender for Cloud Event Collector provides unified security management and advanced threat protection across hybrid cloud workloads.
```

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector_description.md [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+| API / Permissions name | Type | Description |
+| --- | --- | --- |
+| Azure scurity center | --- | --- |
+| user_impresonation | Delegated | Access Azure security center |
+| SecurityCenter | --- | --- |
+| user_impresonation | Delegated | Access SecurityCenter |
+
+After you finish configuring your application with the proper permission,
+add a “Security Reader” role to the application from the subscription.
+In order to add a role to a subscription refer to:
```
Comment (ShirleyDenkberg) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector_description.md:21]:
```suggestion
In order to add a role to a subscription, refer to:
```

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector_description.md [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+* [Azure AD built-in roles](https://learn.microsoft.com/en-us/azure/active-directory/roles/permissions-reference)
+
+#### Additional information
+
+For additional information on the API refer to:
```
Comment (ShirleyDenkberg) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/MicrosoftDefenderForCloudEventCollector_description.md:26]:
```suggestion
For additional information on the API, refer to:
```

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+    | Subscription ID to use |  | True |
+    | First fetch time interval | First time to start fetching alerts from. | False |
+    | Trust any certificate (not secure) |  | False |
+    | Use system proxy settings |  | False |
+
+4. Click **Test** to validate the URLs, token, and connection.
+
+## Commands
+
+You can execute these commands from the Cortex XSOAR CLI, as part of an automation, or in a playbook.
```
Comment (ShirleyDenkberg) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md:26]:
```suggestion
You can execute these commands from the Cortex XSIAM CLI, as part of an automation, or in a playbook.
```

File: Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+After you successfully execute a command, a DBot message appears in the War Room with the command details.
+
+## Known limitations
+This integration does not have a fetch limit parameter due to the limitations of the API functionality.
+1. The collector fetches all events between the current time and the last time it was fetched during every fetch operation.
+2. If the command is run for the first time, all events from ***first_fetch*** to now will be fetched in one execution.
+It is possible that the cause of 1 and 2 may cause the fetch to take some time. You may need to increase the collector time out in the server configuration if the collector fetch times out.
```
Comment (ShirleyDenkberg) [inline Packs/AzureSecurityCenter/Integrations/MicrosoftDefenderForCloudEventCollector/README.md:33]:
```suggestion
It is possible that the above limitations may cause the fetch to take some time. You may need to increase the collector time out value in the server configuration if the collector fetch times out.
```

**Comment by ShirleyDenkberg** [comment]
@gal-forer @DeanArbel @yucohen @michal-dagan Doc review completed.

**Pull request merged**